### ISSUE TITLE
Troubles replicating paper results

### ISSUE THREAD

**Issue by twoleggedeye** [VisualComputingInstitute/triplet-reid#11]
Hi, i tried to reproduce your results for my research and got poor performance. I've done the followng:

1. Download market-1501 dataset here http://www.liangzheng.org/Project/project_reid.html
2. Download resnet-50v1 from tf.slim models
3. Clone this repo and execute market1501_train.sh with specified paths and default parameters

In the end of the training i got around 0.15 rank-1 accuracy according to tensorboard logs, and around 0.84 rank-1 at 2000 iterations, see attached screenshot
![tensorboard](https://user-images.githubusercontent.com/13769095/33229501-160fee2c-d1e1-11e7-8e15-365caaada4da.PNG)

Is there any chance that something wrong with default hyperparameters or with reimplementation code?
Thanks!

**Comment by lucasb-eyer** [comment]
That's odd. It seems like something happend around the 2k mark that threw the model off and it's gone downhill since then. I haven't seen logs that look like this yet, and I've re-run the model with `market1501_train.sh` before publishing the code, replicating the paper results. Furthermore, @ergysr told me he was able to reproduce them, too. That suggests something's wrong on your end, but I'm at a loss to what it could be.

You did run the `market1501_train.sh` with only the three required parameters, or did you set some others? Could you post the `args.json` file from your experiment folder?

**Comment by twoleggedeye** [comment]
I specified only paths, without any additional parameters, here is the contents of args.json
```
{
  "batch_k": 4,
  "batch_p": 18,
  "checkpoint_frequency": 1000,
  "crop_augment": true,
  "decay_start_iteration": 15000,
  "detailed_logs": false,
  "embedding_dim": 128,
  "experiment_root": "/mnt/storage/experiments/triplet-reid/",
  "flip_augment": true,
  "head_name": "fc1024",
  "image_root": "/mnt/storage/datasets/Market-1501/",
  "initial_checkpoint": "/mnt/storage/models/resnet_v1_50.ckpt",
  "learning_rate": 0.0003,
  "loading_threads": 8,
  "loss": "batch_hard",
  "margin": "soft",
  "metric": "euclidean",
  "model_name": "resnet_v1_50",
  "net_input_height": 256,
  "net_input_width": 128,
  "pre_crop_height": 288,
  "pre_crop_width": 144,
  "resume": false,
  "train_iterations": 25000,
  "train_set": "data/market1501_train.csv"
}

```

[and zipped json itself](https://github.com/VisualComputingInstitute/triplet-reid/files/1503004/args.zip)

**Comment by lucasb-eyer** [comment]
Thanks. That all looks correct, so I'm really at a loss here. Especially that in your tensorboard screenshot everything looks fine until ~2k iterations where something happens and it breaks. I've never had that with all default parameters, only when I did something wrong in the code. What does the "histograms" tab of tensorboard look like?

The few things I'd try next, but they are all quite far-fetched because I have no idea:
- run memcheck on your GPU, just to be sure.
- run training again, maybe with `--detailed_logs` so that we can have a closer look if it happens again.
- double-check if the dataset got corrupted during download.

**Comment by Pandoro** [comment]
I can only agree with Lucas. This is very curious. I'm additionally wondering what TensorFlow version you are using and if at any point you canceled and resumed the training.

To make sure we have the same data I created md5 checksums for all our images. You could use this to make sure the data is the same.
[market1501_md5sums.txt.zip](https://github.com/VisualComputingInstitute/triplet-reid/files/1503033/market1501_md5sums.txt.zip)

**Comment by twoleggedeye** [comment]
Here is the histograms tab 
![tensorboard_hist](https://user-images.githubusercontent.com/13769095/33230187-cfadc724-d1ef-11e7-8d5f-3a4d6b7240de.png)

Actually, i already restarted training with detailed log, and it looks much better now though it's not done yet:

![tensorboard_restart](https://user-images.githubusercontent.com/13769095/33230219-4f96186a-d1f0-11e7-96cd-d7c47af88147.PNG)

**Comment by twoleggedeye** [comment]
@Pandoro I'm using tensorflow-gpu 1.4 installed from pip, and training wasn't interrupted

**Comment by lucasb-eyer** [comment]
Thanks for all the extra info, the histograms look broken too, probably that huge spike at 2k somehow broke the model. Interesting and weird.

You still have some serious spikes in the second loss curve, which I don't think I had, so there's still something weird. I've just added a screenshot of a healthy training to the README, it still looks different.

**Comment by lucasb-eyer** [comment]
Hi, I believe I have found the cause of your problems!

I have introduced a very stupid bug [here](https://github.com/VisualComputingInstitute/triplet-reid/blob/master/loss.py#L47-L50) while cleaning up the code. I effectively switched `euclidean` and `sqeuclidean`, meaning you have trained with square-euclidean distance so far, which explains why your training was less stable. (We also experienced instability with squared euclidean.)

I'm currently waiting for a training to finish and will push the fix in a couple hours when the training is done and I have verified results.
